### ISSUE TITLE
Brightness issue for HDR/HDR10 prints in default Stremio player on Mac OS.

### ISSUE THREAD

**Issue by gunpreet34** [Stremio/stremio-bugs#646]
TL;DR:
The Stremio default player is showing very less brightness on Macbook pro 2021 (and a lot of other HDR supported devices - left for another issue). Even with 1600 nits peak brightness, some scenes are very dimly lit in HDR prints where SDR prints shines (but then again we don't have 4K SDR prints out there for every content).

Screenshots attached:
1. [External player IINA](https://github.com/Stremio/stremio-bugs/assets/22132858/efed1545-d32b-41eb-9505-afa6a16381f8)
2. [Default stremio player](https://github.com/Stremio/stremio-bugs/assets/22132858/8cad2b3b-749e-4a24-be93-ee1959d7f9da)

I have been seeing pretty consistently and the issue is 100% reproducible. Steps to reproduce:
The setup I'm using is Stremio+Torrentio+RD
1. Find any media with HDR/HDR10+ print (Wheel of Time Season2 has almost all prints available in 4k).
2. Play the media in default Stremio player.
3. Copy link url from bottom corner and stream the same episode/movie in IINA player.

PS: I'm also seeing pink and green colors in Dolby Vision files but I think that's due to decoding of Macbook and that there's no player out there that could support the DV prints. Source: [Reddit](https://www.reddit.com/r/mac/comments/oylocp/dolby_vision_mkv_on_macbook_pro_16/).
If you want me to, I can open another bug for the Dolby Vision thingy.

Hoping to get the issue fixed soon, thanks for the fabulous player.

**Comment by gunpreet34** [comment]
To even confirm it's HDR only problem and not a setting in IINA player, you could try streaming the 4K HDR vs 4K SDR print (again, Wheel of Time S02E01 has both versions available). Scene at 9:36 with the teardrop is very suitable for comparison.

**Comment by TRtomasz** [comment]
check if you have hardware acceleartion turned on in the settings if not try turning it on

**Comment by kokroo** [comment]
@gunpreet34 @TRtomasz 
I can confirm that Stremio on Apple Silicon Macs does not support HDR inside its player. This is not new information, you can find hundreds of comments on Reddit, etc. The hardware acceleration switch makes no difference to the HDR capability at least on macOS.

Stremio code needs to be updated/fixed to enable HDR playback. For example Infuse Player on macOS can successfully play HDR content from files.

**Comment by TRtomasz** [comment]
@kokroo  We do not control the player we use MPV, if you can check and confirm that you have HDR with MPV on macos then perhaps we can have it too.

**Comment by kokroo** [comment]
> @kokroo We do not control the player we use MPV, if you can check and confirm that you have HDR with MPV on macos then perhaps we can have it too.

I can't even find an Apple Silicon native version of MPV for macOS to test with! Reading up on Google, I see a lot of people asking about HDR support on mpv but being pointed to IINA player by others.

Is MPV just a lib, or a full-blown app with a GUI?

**Comment by TRtomasz** [comment]
They have an app too https://mpv.io/

**Comment by kokroo** [comment]
@TRtomasz https://github.com/mpv-player/mpv/issues/7341

The MPV HDR support for macOS has been added, as far as I can tell. Can we upstream the updates into the next Stremio version?

**Comment by Akemi** [comment]
if you are using libmpv and the render API HDR/EDR support is completely independent from the support in mpv. you have to implement it yourself by setting the correct color space on your CAOpenGLLayer yourself. that was always the case.

also this only adds HDR/EDR support on the old cocoa-cb backend. it was already supported on the new macvk/vulkan/gpu(-next) backendend before.

as an example see [iina](https://github.com/iina/iina/pull/3701/files).

**Comment by kokroo** [comment]
> if you are using libmpv and the render API HDR/EDR support is completely independent from the support in mpv. you have to implement it yourself by setting the correct color space on your CAOpenGLLayer yourself. that was always the case.
> 
> also this only adds HDR/EDR support on the old cocoa-cb backend. it was already supported on the new macvk/vulkan/gpu(-next) backendend before.
> 
> as an example see [iina](https://github.com/iina/iina/pull/3701/files).

So what would we need to change in Stremio to enable this now?

**Comment by Akemi** [comment]
that question is quite broad isn't? i have no insight of stremio specifically and can only give a general direction. the linked iina PR is already a good indication what needs to be done. you could also look at the [commit](https://github.com/mpv-player/mpv/commit/8a61929eb8e62097cf8ef0764dba233000fd1e5f) that added support to mpv. though in the case of mpv those are options that need to be manually set, preferably to the color primaries/transfer characteristics of the VO target.

TLDR:
- set `wantsExtendedDynamicRangeContent = true` on the CAOpenGLLayer
- set `wantsExtendedDynamicRangeOpenGLSurface = true` on the view containing the above layer
- set the proper [colour space](https://developer.apple.com/documentation/coregraphics/cgcolorspace#1692873) on the above layer (`layer.colorspace =`), to the color primaries/transfer characteristics of the VO target (see below ``--target-trc`` and ``--target-prim``)
- set primaries/transfer characteristics ``--target-trc`` and ``--target-prim`` of the display

first 3 make macOS aware that the window/view/layer is presenting something in HDR/EDR, 3 alone tells macOS what colour space the content is rendered in, 4 tells mpv to what characteristics to render to.

also see the [Apple docs](https://developer.apple.com/documentation/metal/hdr_content/displaying_hdr_content_in_a_metal_layer) for the 3 possible ways as reference. they removed the opengl ones sadly, so one has to transfer a little bit.

**Comment by kprabhat248** [comment]
Also the desktop version is quite old waiting for update just like on 1.6.12 on arm

**Comment by AlvinHV** [comment]
HDR support will be available for macOS in a native app instead of the desktop version. Playing with AVPlayer will likely support HDR, and VLCKit is most likely to support HDR as well.

**Comment by kokroo** [comment]
> HDR support will be available for macOS in a native app instead of the desktop version. 

Is this something that has been planned, or is this just a possibility at this point?

**Comment by AlvinHV** [comment]
> > HDR support will be available for macOS in a native app instead of the desktop version. 
> 
> 
> 
> Is this something that has been planned, or is this just a possibility at this point?

iOS app definitely coming soon. MacOS port will definitely come from that but we are not sure how it will. When iOS app releases on App store apple silicon macs would be able to run it(preferably as macCatalyst). We still going to maintain previous desktop app as app store app for mac is lowest priority right now and near future

**Comment by kokroo** [comment]
> iOS app definitely coming soon. MacOS port will definitely come from that

@AlvinHV But the iOS app does not support torrent sources. There is a lot of free and open media on torrents, such as old movies and copyright-free films. If the macOS app will just be another version of the iOS app, then we are cooked.

**Comment by AlvinHV** [comment]
@kokroo You have to differentiate between iOS operating system and Appstore guidelines. Torrent feature is not technical limitation at all. Right now you can run iOS app with stremio server app to achieve torrent feature so that way torrenting would happen on stremio server. Additionally, macOS app can be distributed directly from our web site totally avoiding apple's App Store limitations. For now new stremio macOS app would be webkit app similar to recent linux and windows release. Right now I'm focused on tvOS and iOS.

**Comment by kokroo** [comment]
@AlvinHV Perhaps we could implement HDR in the present macOS app?
I am trying to create a PR for enabling native Apple Silicon builds as well.

**Comment by AlvinHV** [comment]
@kokroo well problem is player uses OpenGL to present video but OpenGL does not support HDR colors so it would look bad. Implementing HDR requires player that uses [AVSampleBufferDisplayLayer](https://developer.apple.com/documentation/avfoundation/avsamplebufferdisplaylayer) and Metal layer. For that huge rewrite needed for existing macOS app so we could implement apple native players like KSPlayer

**Comment by watch2together** [comment]
@kokroo 
>But the iOS app does not support torrent sources. There is a lot of free 
>and open media on torrents, such as old movies and copyright-free films. 
>If the macOS app will just be another version of the iOS app, then we are cooked.

We use debrid.  Is your internet free at home or free on your phone?  Was your phone free?  Was your laptop free?  Why the resistance to not-free debrid?

Your storage device was not-free.  P2P thrashing raises your replacement frequency and hence your costs.

We are disappointed to have missed Stremio Lite for ios.  That was enough functionality with [DMM](https://github.com/debridmediamanager/debrid-media-manager)

>Stremio code needs to be updated/fixed to enable HDR playback.

it's a tad more complex that that.  Sink-led vs player-led

@AlvinHV 
> HDR support will be available for macOS in a native app instead of 
>the desktop version. Playing with AVPlayer will likely support HDR, and 
> VLCKit is most likely to support HDR as well.

Awesome news for VLC

**Comment by Albabit** [comment]
Hi !
I still have the same issue here.
You can see on the screenshot attached, result with IINA player on the left, and default Stremio player on the right.
- Dolby Vision Content on Macbook Pro M1, Mac-OS 26.3.
- Stremio App 5.1.15.

![Image](https://github.com/user-attachments/assets/d5fa53a5-9aa4-4b73-9580-c373f827afa1)

Also, in the Stremio settings, if I select "IINA" player as external player, rather than the default player, when I launch a content :
I get :
"Stream opened in external player"
+
"LibraryltemsPushedToAPI
Failed to fetch: Load failed".
Messages.

![Image](https://github.com/user-attachments/assets/29280c5c-be2f-4fc6-b9e3-ac2a544e9c4a)

The first time.
Then, only : "Stream opened in external player".
But nothing is opened in IINA player.

**Comment by kprabhat248** [comment]
> Hi ! I still have the same issue here. You can see on the screenshot attached, result with IINA player on the left, and default Stremio player on the right.
> 
> * Dolby Vision Content on Macbook Pro M1, Mac-OS 26.3.
> * Stremio App 5.1.15.
> 
> ![Image](https://github.com/user-attachments/assets/d5fa53a5-9aa4-4b73-9580-c373f827afa1)
> 
> Also, in the Stremio settings, if I select "IINA" player as external player, rather than the default player, when I launch a content : I get : "Stream opened in external player" + "LibraryltemsPushedToAPI Failed to fetch: Load failed". Messages.
> 
> ![Image](https://github.com/user-attachments/assets/29280c5c-be2f-4fc6-b9e3-ac2a544e9c4a)
> 
> The first time. Then, only : "Stream opened in external player". But nothing is opened in IINA player.

 I think recent update  broke external player as iina hope infuse get added as external player on desktop version too

**Comment by Albabit** [comment]
> I think recent update broke external player as iina hope infuse get added as external player on desktop version too

I'm not able to test if the problem is the same with previous version 4.4.175, because there is no option for external player in this version.

**Comment by kprabhat248** [comment]
> > I think recent update broke external player as iina hope infuse get added as external player on desktop version too
> 
> I'm not able to test if the problem is the same with previous version 4.4.175, because there is no option for external player in this version.

I think iina was broken on the v5 beta version desktop one from release others work well I raised it before too I simple copy stream link and paste in open url for iina others work fine vlc and mpv too

**Comment by Albabit** [comment]
> I think iina was broken on the v5 beta version desktop one from release others work well I raised it before too I simple copy stream link and paste in open url for iina others work fine vlc and mpv too

Ok, so let's hope it'll be fixed on future releases of v.5 :)
I confirm copy stream link to IINA is working, it's not the problem here ;)
